### PR TITLE
Hw fruit shop, clean pull request

### PR DESCRIPTION
**Pull request by Likh-Alex**

### CODE REVIEW

File: src/main/java/service/DataWriter.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package service;
+
+import java.util.Map;
+import model.Fruit;
+
+public interface DataWriter {
+    void writeToFile(Map<Fruit, Integer> fruitReport, String fileTo);
```
Comment (Ketch92) [inline src/main/java/service/DataWriter.java:7]: fileTo is a file or a path to file?

File: src/main/java/service/implementation/CsvDataParser.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        String[] parsedLine = line.split(COMMA);
+        try {
+            return new FruitTransactionDto(Operation.fromString(parsedLine[OPERATION]),
+                    new Fruit(parsedLine[FRUIT_NAME].toLowerCase()),
+                    Integer.parseInt(parsedLine[FRUITS_AMOUNT]));
+        } catch (NumberFormatException nfe) {
+            throw new InvalidDataFormatException("Input data has invalid format: "
+                    + Arrays.toString(parsedLine) + "\n"
+                    + "should be ['o' - operation, 'fruit', '3'- amount]");
+        }
```
Comment (Ketch92) [inline src/main/java/service/implementation/CsvDataParser.java:27]: I don't know if it's ok to catch NumberFromatException

File: src/main/java/service/implementation/FruitShopServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Override
+    public void applyOperationsOnFruitsDto(List<FruitTransactionDto> fruitTransactionDtoList) {
+        for (FruitTransactionDto transactionDto : fruitTransactionDtoList) {
+            try {
+                Operation operation = transactionDto.getOperation();
+                operationStrategyMap.get(operation).apply(transactionDto);
+            } catch (RuntimeException e) {
+                throw new InvalidDataFormatException("Your data has invalid format,"
+                        + " something is missing" + "\n" + e);
```
Comment (Ketch92) [inline src/main/java/service/implementation/FruitShopServiceImpl.java:28]: you can declare a contstant field with exception message

File: src/test/java/model/OperationTest.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package model;
+
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+
+public class OperationTest {
+    private static final String INVALID_OPERATION = "?";
```
Comment (Ketch92) [inline src/test/java/model/OperationTest.java:8]: no need to test enum, add it to exclude of jacoco

**Review by Ketch92 — CHANGES_REQUESTED** [review]

File: src/main/java/service/DataWriter.java (hunk 1 above)
Comment (Likh-Alex) [reply src/main/java/service/DataWriter.java]: for clarification - renamed it to the `fileName`

File: src/main/java/service/implementation/CsvDataParser.java (hunk 2 above)
Comment (Likh-Alex) [reply src/main/java/service/implementation/CsvDataParser.java]: changed to catch a `RuntimeException`

File: src/main/java/service/implementation/FruitShopServiceImpl.java (hunk 3 above)
Comment (Likh-Alex) [reply src/main/java/service/implementation/FruitShopServiceImpl.java]: Done - now it's clearer)

File: src/test/java/model/OperationTest.java (hunk 4 above)
Comment (Likh-Alex) [reply src/test/java/model/OperationTest.java:8]: added to `exclude` in pom.xml

**Review by Ketch92 — APPROVED** [review]

File: src/main/java/model/Operation.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    public static Operation fromString(String stringType) {
+        for (Operation type : Operation.values()) {
+            if (type.operation.equals(stringType)) {
+                return type;
+            }
+        }
+        throw new IllegalArgumentException("You have entered invalid operation: "
+                + stringType + "." + "\nValid operations: "
+                + Arrays.toString(Operation.values()));
```
Comment (dmytro-hryhoruk) [inline src/main/java/model/Operation.java:29]:
```suggestion
        throw new IllegalArgumentException("You have entered invalid operation: "
                + stringType + ".\nValid operations: "
                + Arrays.toString(Operation.values()));
```
Better use String.format()

File: src/main/java/service/implementation/CsvDataParser.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    @Override
+    public FruitTransactionDto parse(String line) throws InvalidDataFormatException {
+        String[] parsedLine = line.split(COMMA);
+        try {
+            return new FruitTransactionDto(Operation.fromString(parsedLine[OPERATION]),
+                    new Fruit(parsedLine[FRUIT_NAME].toLowerCase()),
+                    Integer.parseInt(parsedLine[FRUITS_AMOUNT]));
+        } catch (RuntimeException e) {
+            throw new InvalidDataFormatException("Input data has invalid format: "
+                    + Arrays.toString(parsedLine) + "\n"
```
Comment (dmytro-hryhoruk) [inline src/main/java/service/implementation/CsvDataParser.java:25]: sane about String.format()

File: src/test/java/service/implementation/CsvDataReaderImplTest.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    public static final String VALID_INPUT = "src/test/resources/valid_input.csv";
+    public static final String INVALID_AMOUNT_NAN = "src/test/resources/invalid_amount_NaN.csv";
+    public static final String INVALID_FORMAT = "src/test/resources/invalid_format.csv";
+    public static final String INVALID_OPERATION_IS_NUMBER = "src/test/resources/"
+            + "invalid_operation_isNumber.csv";
+
+    private static List<FruitTransactionDto> expectedFruits;
+    private static DataReader dataReader;
+
+    @Before
```
Comment (dmytro-hryhoruk) [inline src/test/java/service/implementation/CsvDataReaderImplTest.java:27]: Do you need to create these data before each test method?

File: src/test/java/service/implementation/CsvDataWriterImplTest.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+import service.DataWriter;
+
+public class CsvDataWriterImplTest {
+    public static final String CORRECT_OUTPUT = "src/test/resources/correct_output.csv";
+    public static final String EXPECTED_OUTPUT = "src/test/resources/expected_output.csv";
+    private static final Map<Fruit, Integer> testFruits = new HashMap<>();
+    private DataWriter dataWriter;
+
+    @Before
+    public void setUp() {
```
Comment (dmytro-hryhoruk) [inline src/test/java/service/implementation/CsvDataWriterImplTest.java:25]: same

**Review by dmytro-hryhoruk — APPROVED** [review]
Good job

File: src/main/java/model/Operation.java (hunk 5 above)
Comment (Likh-Alex) [reply src/main/java/model/Operation.java]: Done, refactored to display message with `String.format()`. Thank you for the note

File: src/main/java/service/implementation/CsvDataParser.java (hunk 6 above)
Comment (Likh-Alex) [reply src/main/java/service/implementation/CsvDataParser.java]: Done here as well

File: src/test/java/service/implementation/CsvDataReaderImplTest.java (hunk 7 above)
Comment (Likh-Alex) [reply src/test/java/service/implementation/CsvDataReaderImplTest.java:26]: I thought it is good to initialise Object in the `Before` method, I see that it does not make sense workflow wise. Thank you

File: src/test/java/service/implementation/CsvDataWriterImplTest.java (hunk 8 above)
Comment (Likh-Alex) [reply src/test/java/service/implementation/CsvDataWriterImplTest.java:25]: Done here as well

**Pull request closed without merging**